OSDOCS#11115: Add OLMv1 object ownership docs (#82517)
openshift/openshift-docs: modules/olmv1-object-ownership.adoc

// Module included in the following assemblies:
//
// * extensions/arch/operator-controller.adoc

:_mod-docs-content-type: CONCEPT

[id="olmv1-object-ownership_{context}"]
= Object ownership for cluster extensions

> In {olmv1-first}, a Kubernetes object can only be owned by a single `ClusterExtension` object at a time. This ensures that objects within a {product-title} cluster are managed consistently and prevents conflicts between multiple cluster extensions attempting to control the same object.

Reviewer: ```suggestion
In {olmv1-first}, a Kubernetes object can only be owned by a single `ClusterExtension` object at a time. This ensures that objects within an {product-title} cluster are managed consistently and prevents conflicts between multiple cluster extensions attempting to control the same object.
```

[id="olmv1-single-ownership_{context}"]
== Single ownership

The core ownership principle enforced by {olmv1} is that each object can only have one cluster extension as its owner. This prevents overlapping or conflicting management by multiple cluster extensions, ensuring that each object is uniquely associated with only one bundle.

.Implications of single ownership

> 1. Bundles that provide a CRD can only installed once.

Reviewer: ```suggestion
1. Bundles that provide a CRD can only be installed once.
```

+
Bundles provide `CustomResourceDefinition` objects (CRDs), which are part of a `ClusterExtension` object. This means a bundle can only be installed once in a cluster. Attempting to install another bundle that provides the same CRD will result in failure, as each custom resource can have only one cluster extension as its owner.

2. Cluster extensions cannot share objects.
+
{olmv1}'s single-owner policy means that cluster extensions cannot share ownership of any objects. If one cluster extension manages a specific obect, such as a `Deployment`, `CustomResourceDefinition`, or `Service` object, another cluster extension cannot claim ownership of the same object. Any attempt to do so will be blocked by {olmv1}.

[id="olmv1-error-messages_{context}"]
== Error messages

> When a conflict occurs due to multiple cluster extensions attempting to manage the same object, Operator Controller returns a clear error message, indicating the ownership conflict:

Reviewer 2: ```suggestion
When a conflict occurs due to multiple cluster extensions attempting to manage the same object, Operator Controller returns the following error message, indicating the ownership conflict:
```

.Example error message
[source,text]
----
CustomResourceDefinition 'logfilemetricexporters.logging.kubernetes.io' already exists in namespace 'kubernetes-logging' and cannot be managed by operator-controller
----

This error message signals that the object is already being managed by another cluster extension and cannot be reassigned or shared.

[id="olmv1-ownership-considerations_{context}"]
== Considerations

> As a cluster or extension administrator, keep the following considerations in mind:

Reviewer 2: ```suggestion
As a cluster or extension administrator, review the following considerations:
```

Uniqueness of bundles::

> Ensure that operator bundles providing the same CRDs are not installed more than once. This can prevent potential installation failures due to ownership conflicts.

Reviewer 2: ```suggestion
Ensure that Operator bundles providing the same CRDs are not installed more than once. This can prevent potential installation failures due to ownership conflicts.
```

> Avoid object sharing:: If you require different cluster extensions to interact with similar resources, ensure they are managing separate objects. Cluster extensions cannot jointly manage the same object due to the single-owner enforcement.

Reviewer 2: ```suggestion
Avoid object sharing:: If you need different cluster extensions to interact with similar resources, ensure they are managing separate objects. Cluster extensions cannot jointly manage the same object due to the single-owner enforcement.
```
